http/http_client.h

```diff
@@ -0,0 +1,432 @@
+                                    queryParams, headers, timeout);
+    }
+
+    size_t queuedRequests()
+        const
+    {
+        return impl->queuedRequests();
+    }
+
+    HttpClient & operator = (HttpClient && other) noexcept
+    {
+        if (&other != this) {
+            impl = std::move(other.impl);
+        }
+
+        return *this;
+    }
+
+private:
+    std::unique_ptr<HttpClientImpl> impl;
+};
+
+
+/****************************************************************************/
+/* HTTP CLIENT CALLBACKS                                                    */
+/****************************************************************************/
+
+struct HttpClientCallbacks {
+    typedef std::function<void (const HttpRequest &,
+                                const std::string &,
+                                int code)> OnResponseStart;
+    typedef std::function<void (const HttpRequest &,
+                                const char * data, size_t size)> OnData;
+    typedef std::function<void (const HttpRequest & rq,
+                                HttpClientError errorCode)> OnDone;
+
+    HttpClientCallbacks(OnResponseStart onResponseStart = nullptr,
+                        OnData onHeader = nullptr,
+                        OnData onData = nullptr,
+                        OnDone onDone = nullptr)
+        : onResponseStart_(onResponseStart),
+          onHeader_(onHeader), onData_(onData),
+          onDone_(onDone)
+    {
+    }
+
+    virtual ~HttpClientCallbacks()
+    {
+    }
+
+    static const std::string & errorMessage(HttpClientError errorCode);
+
+    /* initiates a response */
+    virtual void onResponseStart(const HttpRequest & rq,
+                                 const std::string & httpVersion,
+                                 int code);
+
+    /* callback for header lines, one invocation per line */
+    virtual void onHeader(const HttpRequest & rq,
+                          const char * data, size_t size);
+
+    /* callback for body data, one invocation per chunk */
+    virtual void onData(const HttpRequest & rq,
+                        const char * data, size_t size);
+
+    /* callback for operation completions, implying that no other call will
+     * be performed for the same request */
+    virtual void onDone(const HttpRequest & rq,
+                        HttpClientError errorCode);
+
+private:
+    OnResponseStart onResponseStart_;
+    OnData onHeader_;
+    OnData onData_;
+    OnDone onDone_;
+};
+
+
+/****************************************************************************/
+/* HTTP CLIENT SIMPLE CALLBACKS                                             */
+/****************************************************************************/
+
+/* This class is a child of HttpClientCallbacks and offers a simplified
+ * interface when support for progressive responses is not necessary. */
+
+struct HttpClientSimpleCallbacks : public HttpClientCallbacks
```

Comment:
I could be it does mix both point of views in the same class, is that desirable?